crates/tx5-go-pion/src/peer_con.rs

```diff
@@ -200,14 +202,50 @@ impl PeerConnection {
         .await?
     }
 
+    /// Set the connection state. This should only be set based on connection state events
+    /// coming from the underlying webrtc library.
+    pub fn set_con_state(&self, con_state: PeerConnectionState) {
+        let mut lock = self.0.lock().unwrap();
+        if let Ok(core) = &mut *lock {
+            core.con_state = con_state;
+        } else {
+            tracing::warn!(
+                ?con_state,
+                "Unable to set peer connection state: {:?}",
+                self.get_peer_con_id()
+            );
+        }
+    }
+
+    /// Get the connection state.
+    pub fn get_con_state(&self) -> Result<PeerConnectionState> {
+        peer_con_strong_core!(self.0, core, { Ok(core.con_state) })
+    }
+
     /// Close this connection.
-    pub fn close(&self, err: Error) {
+    pub fn close<E: Into<Error>>(&self, err: E) {
+        let err = err.into();
         let mut tmp = Err(err.clone());
 
         {
             let mut lock = self.0.lock().unwrap();
             let mut do_swap = false;
             if let Ok(core) = &mut *lock {
+                // Connections should only be getting closed when they have been established or have failed
+                // to establish. This function is effectively a 'force close' and is expected to be called with
+                // with an error. Allow the close anyway but warn the caller.
+                match core.con_state {
+                    PeerConnectionState::Connected
+                    | PeerConnectionState::Failed => {
+                        // Allow the connection to be closed
+                    }
+                    _ => {
+                        tracing::warn!(
+                            "Closing peer connection in invalid state: {:?}",
+                            core.con_state
+                        );
```

Comment:
Why can they not be closed while being established? In the case where two peers simultaneously send offers to each other, one will establish a peerconnection that gets overridden, so this close will be called on that during normal operation.

Reply:
I was reading something about not closing connections while they're being established but it looks like that was specific to another implementation in a browser context, so no idea if it applies here. It sounds like it doesn't so I'll remove this code :)

Reply:
I read something about problems with doing so but that was about a different implementation so maybe not relevant here. I'll get rid of this stuff

Reply:
This does mean that `con_state` isn't used yet and `get_con_state` isn't visible outside the crate. I think that's okay for now, it was already an incomplete code path that is now just a little more implemented